src/datasets/arrow_dataset.py

```diff
@@ -225,117 +235,111 @@ def _get_output_signature(dataset: "Dataset", collate_fn: Callable, collate_fn_a
-                    shape = signatures["input_ids"].shape
+            np_arrays = []
+            for array in raw_arrays:
+                if isinstance(array, np.ndarray):
+                    np_arrays.append(array)
+                elif isinstance(array, tf.Tensor):
+                    np_arrays.append(array.numpy())
                 else:
-                    # If this doesn't look like LM labels that got added by the collate_fn, let's not say anything
-                    # about the dimensions we're unsure of
-                    shape = [batch_size] + [None for dim in tensor.shape.as_list()[1:]]
+                    np_arrays.append(np.array(array))
+
+            if np.issubdtype(np_arrays[0].dtype, np.integer) or np_arrays[0].dtype == np.bool:
+                tf_dtype = tf.int64
+                np_dtype = np.int64
+            elif np.issubdtype(np_arrays[0].dtype, np.number):
+                tf_dtype = tf.float32
+                np_dtype = np.float32
+            elif np_arrays[0].dtype.kind == "U":  # Unicode strings
+                np_dtype = np.unicode_
+                tf_dtype = tf.string
             else:
-                # If this doesn't look like LM labels that got added by the collate_fn, let's not say anything
-                # about the dimensions we're unsure of
-                shape = [batch_size] + [None for dim in tensor.shape.as_list()[1:]]
-            signatures[column] = tf.TensorSpec(shape=shape, dtype=tensor.dtype)
-        return columns_to_dtypes, signatures
+                raise RuntimeError(f"Unrecognized array dtype {np_arrays[0].dtype}!")
+            shapes = [array.shape for array in np_arrays]
+            static_shape = []
+            for dim in range(len(shapes[0])):
+                sizes = set([shape[dim] for shape in shapes])
+                if dim == 0:
+                    static_shape.append(batch_size)
+                    continue
+                if len(sizes) == 1:  # This dimension looks constant
+                    static_shape.append(sizes.pop())
+                else:  # Use None for variable dimensions
+                    static_shape.append(None)
+            tf_columns_to_signatures[column] = tf.TensorSpec(shape=static_shape, dtype=tf_dtype)
+            np_columns_to_dtypes[column] = np_dtype
+        # endregion
+
+        return tf_columns_to_signatures, np_columns_to_dtypes
 
     def to_tf_dataset(
         self,
-        columns: Union[str, List[str]],
         batch_size: int,
-        shuffle: bool,
-        collate_fn: Callable,
-        drop_remainder: bool = None,
-        collate_fn_args: Dict[str, Any] = None,
-        label_cols: Union[str, List[str]] = None,
-        dummy_labels: bool = False,
+        columns: Optional[Union[str, List[str]]] = None,
+        shuffle: bool = True,
+        collate_fn: Optional[Callable] = None,
+        drop_remainder: Optional[bool] = None,
```

Comment:
I think it should be False by default, as `tf.data.Dataset.batch` does.

Reply:
If `shuffle` is `False` by default then `drop_remainder` will also follow it, because the default setting for `drop_remainder` is to use the same value as `shuffle`. The reason for that is that you usually only shuffle the training set, which is also the dataset where dropping the remainder batch is common.

Reply:
Again by default we shouldn't so extra actions rather than what the user asked for, i.e. convert the dataset to a tf dataset.

Even tf.data.Dataset doesn't drop the remainder by default:
https://www.tensorflow.org/api_docs/python/tf/data/Dataset?version=nightly#batch

Your concern makes total sense but is specific to one use case. Feel free to explain it in the docstring though

Reply:
That's fair! We might encounter some backward-compatibility problems if we changed it now, though.

I'm hoping we can move a lot of the more `transformers`-specific stuff to a method in that library, which would become the 'standard' conversion method for users. At that point, we could make these compatibility-breaking changes to `to_tf_dataset()` and make it more like the pure function it should have been in the beginning. Do you think that's okay?

Reply:
Which backward compatibility problems ? Is there a place somewhere that requires XLA and expects to_tf_dataset to ignore some data just for XLA ?

Reply:
Keras uses XLA whenever `jit_compile=True`, which a lot of users use by default, so we might see some issues there. I haven't tested it in detail - do you want me to run some tests, or are you okay leaving it unchanged for now?

Reply:
It goes against the main rationale for this function so yea I'd like to discuss it more. I don't think testing really matters - XLA users should know that they should drop the remainder and asks for this explicitly since this is goes beyond a simple conversion from HF dataset to TF dataset.

My main concern now is is there a place somewhere when we say something like "use to_tf_dataset() without specifying drop_remainder=True explicitly and you can expect the best performance using XLA". Because if there is, this is an issue because it means this function has a bad API. And if not, then we can consider making the behavior of `drop_remainder` consistent with the rationale of this function

Reply:
In general, we don't refer to XLA much in the docs, so I don't think we say anything like that explicitly. However, if we change this then users whose code has `jit_compile=True` may suddenly stop working after the patch. I'm willing to make the change if you think it's necessary, though!

Reply:
After testing, I think most cases still work okay with `drop_remainder` defaulting to False, so I don't object to changing it anymore!